### PR TITLE
[LoRA] feat: lora support for SANA.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

- [x] LoRA support to SANA
- [x] LoRA fine-tuning script
- [ ] Tests for the training script
- [x] README for the training script

Example LoRA fine-tuning command:

```bash
CUDA_VISIBLE_DEVICES=0 accelerate launch train_dreambooth_lora_sana.py \
  --pretrained_model_name_or_path=Efficient-Large-Model/Sana_1600M_1024px_diffusers \
  --dataset_name=Norod78/Yarn-art-style --instance_prompt="a puppy, yarn art style" \
  --output_dir=yarn_art_lora_sana \
  --mixed_precision=bf16 --use_8bit_adam \
  --weighting_scheme=none \
  --resolution=1024 --train_batch_size=1 --repeats=1 \
  --learning_rate=1e-4 --report_to=wandb \
  --gradient_accumulation_steps=1 --gradient_checkpointing \
  --lr_scheduler=constant --lr_warmup_steps=0 --rank=4 \
  --max_train_steps=700 --checkpointing_steps=2000 --seed=0 \
  --validation_prompt="a puppy in a pond, yarn art style" --validation_epochs=1 \
  --push_to_hub
```

## Notes

* `mixed_precision="fp16"` is leading to NaN loss values despite the recommendation to use FP16 for "Efficient-Large-Model/Sana_1600M_1024px_diffusers".
* VAE is kept in FP32, always. 
* When FP16 mixed-precision is used, we cast the LoRA params to FP32 while keeping the transformer in FP16. 
* Only QKV is LoRA targeted for LoRA.

## Results

https://wandb.ai/sayakpaul/dreambooth-sana-lora/runs/tf9fo8o6

image

Pre-trained LoRA: https://huggingface.co/sayakpaul/yarn_art_lora_sana

### CODE REVIEW

File: tests/lora/test_lora_layers_sana.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            "max_sequence_length": sequence_length,
+            "output_type": "np",
+            "complex_human_instruction": None,
+        }
+        if with_generator:
+            pipeline_inputs.update({"generator": generator})
+
+        return noise, input_ids, pipeline_inputs
+
+    @unittest.skip("Not supported in Sana.")
```
Comment (sayakpaul) [inline tests/lora/test_lora_layers_sana.py:115]: Skipped tests are the same as Mochi.

File: tests/lora/test_lora_layers_sana.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        num_channels = 4
+        sizes = (32, 32)
+
+        generator = torch.manual_seed(0)
+        noise = floats_tensor((batch_size, num_channels) + sizes)
+        input_ids = torch.randint(1, sequence_length, size=(batch_size, sequence_length), generator=generator)
+
+        pipeline_inputs = {
+            "prompt": "",
+            "negative_prompt": "",
```
Comment (sayakpaul) [inline tests/lora/test_lora_layers_sana.py:101]: Check this internal thread: 
https://huggingface.slack.com/archives/C065E480NN9/p1734324025408149

File: src/diffusers/pipelines/sana/pipeline_sana.py [hunk 3]
```diff
@@ -554,6 +573,10 @@ def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype
     def guidance_scale(self):
         return self._guidance_scale
 
+    @property
+    def attention_kwargs(self):
+        return self._attention_kwargs
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana.py:578]: I think `attention_kwargs` is a better name but LMK if you think otherwise.

**Comment by a-r-r-o-w** [comment]
So was not requested for review, but saw the latest commit on email notifications about `attention_kwargs` rename to `cross_attention_kwargs`. Is it possible to consistently name this simply `attention_kwargs` across the library?

I had shared the concern in a previous lora refactor PR and [this](https://github.com/huggingface/diffusers/blob/3bf5400a64c847e070f332aed7d7a56d89bb22e3/tests/lora/utils.py#L434) comment. This is because I often find myself having to refer to the documentation for different pipelines instead of just being able to use one consistent parameter name to pass lora scale with, and it is frustrating because you wait for the pipeline to load only to find it fail immediately. I'm not sure if others resonate with this, but anyone using loras often will have faced this.

We have `attention_kwargs`, `cross_attention_kwargs` and `joint_attention_kwargs`. IMO these can all just be called `attention_kwargs` and we don't need to make this distinction. I know we can't just change it for all the old pipelines until a release where we are okay with backwards-breaking (1.0.0), but going forward, let's maybe try using same name everywhere? WDYT?

**Comment by sayakpaul** [comment]
Yeah I don't mind.

File: examples/dreambooth/train_dreambooth_lora_sana.py [hunk 4]
```diff
@@ -0,0 +1,1552 @@
...
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        raise ValueError(
+            "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+        )
+
+    # VAE should always be kept in fp32 for SANA (?)
+    vae.to(dtype=torch.float32)
+    transformer.to(accelerator.device, dtype=weight_dtype)
+    # because Gemma2 is particularly suited for bfloat16.
+    text_encoder.to(dtype=torch.bfloat16)
```
Comment (a-r-r-o-w) [inline examples/dreambooth/train_dreambooth_lora_sana.py:994]: I think we could instead load with `torch_dtype=torch.bfloat16` and keept he same comment. This is because weight casting this way ignores `_keep_modules_in_fp32`. I could not get our numerical precision unit tests to match when using the two different ways when working on the integration PR

File: examples/dreambooth/train_dreambooth_lora_sana.py [hunk 5]
```diff
@@ -0,0 +1,1552 @@
...
+        weight_dtype = torch.bfloat16
+
+    if torch.backends.mps.is_available() and weight_dtype == torch.bfloat16:
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        raise ValueError(
+            "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+        )
+
+    # VAE should always be kept in fp32 for SANA (?)
+    vae.to(dtype=torch.float32)
```
Comment (a-r-r-o-w) [inline examples/dreambooth/train_dreambooth_lora_sana.py:991]: FP32 should be good, but I'm not 100% sure. I think AutoencoderDC were all trained in bf16. Maybe @lawrence-cj can comment

File: examples/dreambooth/train_dreambooth_lora_sana.py [hunk 6]
```diff
@@ -0,0 +1,1552 @@
...
+    # because Gemma2 is particularly suited for bfloat16.
+    text_encoder.to(dtype=torch.bfloat16)
+
+    if args.gradient_checkpointing:
+        transformer.enable_gradient_checkpointing()
+
+    if args.lora_layers is not None:
+        target_modules = [layer.strip() for layer in args.lora_layers.split(",")]
+    else:
+        target_modules = ["to_k", "to_q", "to_v"]
```
Comment (a-r-r-o-w) [inline examples/dreambooth/train_dreambooth_lora_sana.py:1002]: We're not finetuning the output layers as we normally do? 👀

File: src/diffusers/loaders/lora_pipeline.py [hunk 7]
```diff
@@ -3254,6 +3254,314 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder"], *
...
+        self.load_lora_into_transformer(
+            state_dict,
+            transformer=getattr(self, self.transformer_name) if not hasattr(self, "transformer") else self.transformer,
+            adapter_name=adapter_name,
+            _pipeline=self,
+            low_cpu_mem_usage=low_cpu_mem_usage,
+        )
+
+    @classmethod
+    # Copied from diffusers.loaders.lora_pipeline.SD3LoraLoaderMixin.load_lora_into_transformer with SD3Transformer2DModel->CogVideoXTransformer3DModel
```
Comment (a-r-r-o-w) [inline src/diffusers/loaders/lora_pipeline.py:3415]:
```suggestion
    # Copied from diffusers.loaders.lora_pipeline.SD3LoraLoaderMixin.load_lora_into_transformer with SD3Transformer2DModel->SanaTransformer2DModel
```

File: src/diffusers/pipelines/pag/pipeline_pag_sana.py [hunk 8]
```diff
@@ -185,6 +189,7 @@ def encode_prompt(
         clean_caption: bool = False,
         max_sequence_length: int = 300,
         complex_human_instruction: Optional[List[str]] = None,
+        lora_scale: Optional[float] = None,
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_sana.py:192]: Are we training text encoder? If not, we can remove these changes maybe

File: src/diffusers/pipelines/pag/pipeline_pag_sana.py [hunk 9]
```diff
@@ -23,15 +23,19 @@
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_sana.py:23]: The PAG pipeline does not support attention kwargs yet

File: src/diffusers/pipelines/sana/pipeline_sana.py [hunk 10]
```diff
@@ -662,6 +686,7 @@ def __call__(
                 [PIL](https://pillow.readthedocs.io/en/stable/): `PIL.Image.Image` or `np.array`.
             return_dict (`bool`, *optional*, defaults to `True`):
                 Whether or not to return a [`~pipelines.stable_diffusion.IFPipelineOutput`] instead of a plain tuple.
+            attention_kwargs: TODO
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/sana/pipeline_sana.py:689]: TODO

**Review by a-r-r-o-w — APPROVED** [review]
Thanks for the super fast work! Looks good to merge after some of the more important reviews are addressed

File: examples/dreambooth/train_dreambooth_lora_sana.py (hunk 5 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sana.py:991]: This is just to be sure VAE's precision isn't a bottleneck for getting good quality training runs. This is anyway a small VAE, won't matter too much I guess.

File: examples/dreambooth/train_dreambooth_lora_sana.py (hunk 6 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sana.py:1002]: Didn't want to start targetting the convs, but can add them, too. Since the output projections are a bit different from the usual, I didn't target them.

File: examples/dreambooth/train_dreambooth_lora_sana.py (hunk 4 above)
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sana.py:994]: Oh this is coming from the example provided in https://huggingface.co/docs/diffusers/main/en/api/pipelines/sana#diffusers.SanaPipeline.__call__.example. In this case, we're doing the exact same thing and we are not fine-tuning the text encoder.

File: src/diffusers/pipelines/pag/pipeline_pag_sana.py (hunk 8 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/pag/pipeline_pag_sana.py:192]: This was to have no surprise for our users when text encoder training support is merged. It's common to see the `encode_prompt()` method being equipped with handling `lora_scale`.

File: tests/lora/test_lora_layers_sana.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+import torch
+from transformers import Gemma2ForCausalLM, GemmaTokenizer
+
+from diffusers import AutoencoderDC, FlowMatchEulerDiscreteScheduler, SanaPipeline, SanaTransformer2DModel
+from diffusers.utils.testing_utils import floats_tensor, is_peft_available, require_peft_backend
+
+
+if is_peft_available():
+    pass
+
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tests/lora/test_lora_layers_sana.py:27]: Urghhh!

File: examples/dreambooth/train_dreambooth_lora_sana.py (hunk 6 above)
Comment (a-r-r-o-w) [reply examples/dreambooth/train_dreambooth_lora_sana.py:1002]: But we only have convs in the VAE. In the transformer, `to_out` is an MLP. Anyway, since it is configurable, it looks good to me

**Comment by sayakpaul** [comment]
@a-r-r-o-w your comments have been addressed. @lawrence-cj could you review / test the training script if you want?

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by sayakpaul** [comment]
Failing tests are unrelated and can safely be ignored. Will add a training test in a followup PR.

**Comment by lawrence-cj** [comment]
> @a-r-r-o-w your comments have been addressed. @lawrence-cj could you review / test the training script if you want?

Working on it. Will fine-tune the model using your pokemon dataset.

File: examples/dreambooth/train_dreambooth_lora_sana.py (hunk 5 above)
Comment (lawrence-cj) [reply examples/dreambooth/train_dreambooth_lora_sana.py:991]: Yes. AutoencoderDC is trained under BF16 and FP32 testing is also fine, just it'll cost a lot of additional GPU memory.

Comment (lawrence-cj) [reply examples/dreambooth/train_dreambooth_lora_sana.py:991]: I mean the VAE.decode() part will consume much more GPU memory if it runs in FP32, specially when the batch_size is more than 1, not the VAE model itself. Not sure if I understand right.

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sana.py:991]: Oh I think we should be good as we barely make use of `decode()` in training.

Comment (lawrence-cj) [reply examples/dreambooth/train_dreambooth_lora_sana.py:991]: OK, that's cool. Then the only concern is that when we visualize the training results during training
